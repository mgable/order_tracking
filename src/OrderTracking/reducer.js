import {
	ORDER_RECEIVED,
	STATUS_RECEIVED,
	RESET_ORDER,
	TIME_RECEIVED,
	SET_THRESHOLD,
	SET_SERVER_STATUS
} from './types';

import { simulationStopped, defaultThreshold, maxOrders, CANCELLED, DELIVERED } from '../config';

// data model
export const initialState = {
	orders: {}, // all currently "active" orders by id
	history: {}, // all "delivered" or "cancelled" orders by id
	currentOrder: null, // last order received
	status: simulationStopped, // simulation status
	time: null, // server time in whole secondes
	threshold: defaultThreshold, // "cooked" filter threshold time
	serverStatus: null // server socket status
};


/** public function
 * @description  model for Order tracking app
 */
const Orders = (state = initialState, action) => {
	switch (action.type) {
		case ORDER_RECEIVED:
			return orderReceived(state, action);
		case STATUS_RECEIVED:
			return setSimulationStatus(state, action);
		case RESET_ORDER:
			return reset(state, action);
		case TIME_RECEIVED:
			return setTime(state, action);
		case SET_THRESHOLD:
			return setThreshold(state, action);
		case SET_SERVER_STATUS:
			return setServerStatus(state, action);
		default:
			return state;
	}
};


const reset = (state, action) => {
	return Object.assign({}, state, {orders: {}, history: {}, time: null, currentOrder: null, status: simulationStopped });
}

const setServerStatus = (state, action) => {
	let serverStatus = action.status;

	if (state.serverStatus !== serverStatus) {
		return Object.assign({}, state, {serverStatus});
	}

	return state;
}

const setThreshold = (state, action) => {
	let threshold = action.threshold;

	if (state.threshold !== threshold){
		return Object.assign({}, state, {threshold});
	}

	return state;
}
 
const setTime = (state, action) => {
	let time = action.time.time;
	if (state.time !== time){
		return Object.assign({}, state, {time});
	}

	return state;
}

const setSimulationStatus = (state, action) => {
	let status = action.status;
	if (state.status !== status){
		return Object.assign({}, state, {status});
	}

	return state;
}

const addToHistory = (state, action) => {
	let currentOrder = action.order,
		id = currentOrder.id,
		history = Object.assign({}, state.history),
		orders = Object.assign({}, state.orders);

	history[id] = currentOrder;

	if (Object.keys(history).length >= maxOrders) { // do not let the history object get too big
		let values = Object.values(history),
			sortedValues = values.sort((v1, v2) => v1.sent_at_second > v2.sent_at_second ? 1 : -1),
			lowestValue = sortedValues.pop(),
			id = lowestValue.id;

		delete history[id];

	}

	delete orders[id];

	return Object.assign({}, state, {currentOrder, history, orders})
}

const addOrder = (state, action) => {
	let currentOrder = action.order;
	if (currentOrder.id) {
		let orders = Object.assign({}, state.orders);

		orders[currentOrder.id] = currentOrder;
		return Object.assign({}, state, {currentOrder, orders});
	}

	return state;
}
 
const orderReceived = (state, action) => {
	let currentOrder = action.order;
	switch (currentOrder.event_name) {
		case CANCELLED: 
		case DELIVERED:
			return addToHistory(state, action);
		default:
			return addOrder(state, action);
	}
};


export default Orders;
