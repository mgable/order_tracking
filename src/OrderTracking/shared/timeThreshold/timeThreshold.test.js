import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, render } from 'enzyme';
import TimeThreshold from './';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import reducer, { initialState }  from '../../reducer';

const mockStore = configureStore(reducer);
const store = mockStore(initialState)

it('renders without crashing', () => {
    const wrapper = render(<Provider store={store}><TimeThreshold  /></Provider>);
  	expect(wrapper).toMatchSnapshot();
});