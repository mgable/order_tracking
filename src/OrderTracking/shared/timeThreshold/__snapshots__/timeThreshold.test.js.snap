// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<span
  class="time-threshold"
>
  <form
    class=""
  >
    <label
      class="form-label"
    >
      please add label
    </label>
    <select
      class="form-control"
    >
      <option
        value="5"
      >
        5
      </option>
      <option
        value="10"
      >
        10
      </option>
      <option
        value="15"
      >
        15
      </option>
      <option
        value="20"
      >
        20
      </option>
    </select>
  </form>
</span>
`;
